src/main/java/org/broadinstitute/hellbender/tools/funcotator/dataSources/vcf/VcfFuncotationFactory.java

```diff
@@ -170,67 +182,108 @@ public String getName() {
+        final Triple<VariantContext, ReferenceContext, List<Feature>> cacheKey = createCacheKey(variant, referenceContext, featureList);
+        final List<Funcotation> cacheResult = cache.get(cacheKey);
+        if (cacheResult != null) {
+            cacheHits++;
+            return cacheResult;
+        }
+
         // Only create annotations if we have data to annotate:
         if ( supportedFieldNames.size() != 0 ) {
 
-            final List<Allele> alternateAlleles = variant.getAlternateAlleles();
-
-            // Create a set to put our annotated Alternate alleles in.
-            // We'll use this to determine if the alt allele has been annotated.
-            final Set<Allele> annotatedAltAlleles = new HashSet<>(alternateAlleles.size());
-
-            if ( !featureList.isEmpty() ) {
-                for ( final Feature feature : featureList ) {
+            // Get rid of any null features.
+            // By this point we know the feature type is correct, so we cast it:
+            final List<VariantContext> funcotationFactoryVariants = featureList.stream().filter(f -> f != null)
+                    .map(f -> (VariantContext) f).collect(Collectors.toList());
 
-                    if ( feature != null ) {
+            // Create a map that will keep the final outputs.  Default it to default funcotations for each alt allele in the
+            //  query variant.
+            final Map<Allele, Funcotation> outputOrderedMap = new LinkedHashMap<>();
+            variant.getAlternateAlleles().forEach(a -> outputOrderedMap.put(a, createDefaultFuncotation(a)));
 
-                        // By this point we know the feature type is correct, so we cast it:
-                        final VariantContext variantFeature = (VariantContext) feature;
+            // TODO: What happens if there is a duplicate pos,ref,alt in the datasource?  See (https://github.com/broadinstitute/gatk/issues/4972)
+            for ( final VariantContext funcotationFactoryVariant : funcotationFactoryVariants ) {
+                final int[] matchIndices = GATKVariantContextUtils.matchAllelesOnly(variant, funcotationFactoryVariant);
 
-                        // Now we create one funcotation for each Alternate allele:
-                        for ( final Allele altAllele : alternateAlleles ) {
-                            if (!(variantFeature.hasAlternateAllele(altAllele) && variantFeature.getReference().equals(variant.getReference()))) {
-                                continue;
-                            }
-                            // Add all Info keys/values to a copy of our default map:
-                            final LinkedHashMap<String, Object> annotations = new LinkedHashMap<>(supportedFieldNamesAndDefaults);
-                            for ( final Map.Entry<String, Object> entry : variantFeature.getAttributes().entrySet() ) {
+                for (int i = 0; i < matchIndices.length; i++) {
```

Comment:
Can you put a comment here stating that `maxIndices.length` will always be equal to the number of alternate alleles in `variant`? 

It's confusing to loop over a different variable's length.

Reply:
Done

Reply:
(The comment is a couple of lines above this)